File: src/day-05-2017.cpp
#include <functional>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

int steps(vector <int> offsets, function <int(int)> mut) {
    auto steps = 0, inc = 0;
    for (auto cur = offsets.begin(); cur < offsets.end() && cur >= offsets.begin(); cur += inc ) {
        steps++;
        inc = *cur;
        *cur = mut(*cur);
    }
    return steps;
}

int main() {
    vector <int> offsets;
    string line;
    while (getline(cin, line)) {
        stringstream val(line);
        int x = 0;
        val >> x;
        offsets.push_back(x);
    }

    // Part 1
    cout << steps(offsets, [](int x){ return x + 1; }) << endl;

    // Part 2
    cout << steps(offsets, [](int x){ return x >= 3 ? x - 1 : x + 1; }) << endl;
}

File: src/perl/day-05-2017.pl
#!/usr/local/bin/perl

use strict;
use warnings;
use 5.010;

sub steps {
    my $oref = shift;
    my @offsets = @{$oref};
    my $mut = shift;

    my $steps = 0;
    my $inc = 0;
    for (my $i = 0; $i <= $#offsets && $i >=0; $i += $inc) {
        $steps++;
        $inc = $offsets[$i];
        $offsets[$i] = $mut->($offsets[$i]);
    }
    return $steps;
}

sub main {
    my @offsets;
    while (<>) {
        chomp;
        push @offsets, $_;
    }

    # Part 1
    say steps(\@offsets, sub { 1 + shift });

    # Part 2
    say steps(\@offsets, sub { my $x = shift; $x >= 3 ? $x - 1 : $x + 1 });
}

exit main();